x/stakeibc/keeper/transfer.go

```diff
@@ -0,0 +1,102 @@
+	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
+	ibctypes "github.com/cosmos/ibc-go/v7/modules/apps/transfer/types"
+	clienttypes "github.com/cosmos/ibc-go/v7/modules/core/02-client/types"
+
+	"github.com/Stride-Labs/stride/v14/x/stakeibc/types"
+)
+
+// Transfers tokens from the community pool deposit ICA account to the host zone holding address for that pool
+func (k Keeper) IBCTransferCommunityPoolTokens(ctx sdk.Context, token sdk.Coin, communityPoolHostZone types.HostZone, shouldLiquidStake bool) error {
+
+	// The memo may contain autopilot commands to atomically liquid stake tokens when transfer succeeds
+	//  both transfer+liquid stake will succeed and stTokens will end in the stride side holding address, 
+	//  or neither will and the original base tokens will return to the foreign deposit ICA address
+	memoCommands := ""
+	autopilotCmd := "{ \"autopilot\": { \"receiver\": \"%s\",  \"stakeibc\": { \"action\": \"LiquidStake\" } } }"
+	if shouldLiquidStake {
+		memoCommands = fmt.Sprintf(autopilotCmd, communityPoolHostZone.CommunityPoolHoldingAddress)
+		stakedDenom := types.StAssetDenomFromHostZoneDenom(token.Denom)
+		k.Logger(ctx).Info(fmt.Sprintf("[IBCTransferCommunityPoolTokens] Transferring %v %s and then liquid staking to %s", token.Amount, token.Denom, stakedDenom))
+	} else {
+		k.Logger(ctx).Info(fmt.Sprintf("[IBCTransferCommunityPoolTokens] Transferring %v %s", token.Amount, token.Denom))
+	}
+	
+	// NOTE: this assumes no clock drift between chains, which tendermint guarantees
+	ibcTransferTimeoutNanos := k.GetParam(ctx, types.KeyIBCTransferTimeoutNanos)
+	timeoutTimestamp := uint64(ctx.BlockTime().UnixNano()) + ibcTransferTimeoutNanos
+
+	msg := ibctypes.NewMsgTransfer(
+		ibctypes.PortID,
+		communityPoolHostZone.TransferChannelId,
+		token,
+		communityPoolHostZone.CommunityPoolDepositIcaAddress, // ICA controlled address on foreign hub
+		communityPoolHostZone.CommunityPoolHoldingAddress, // Stride address, unique to each community pool / hostzone
+		clienttypes.Height{},
+		timeoutTimestamp,
+		memoCommands,
+	)
+
+	// Submit IBC transfer msg
+	msgTransferResponse, err := k.RecordsKeeper.TransferKeeper.Transfer(sdk.WrapSDKContext(ctx), msg)
+	if err != nil {
+		return err
+	}
+
+	k.Logger(ctx).Info(fmt.Sprintf("[IBCTransferCommunityPoolTokens] Successfully submitted ibc transfer message %+v", msgTransferResponse))
+
+	return nil
+}
+
+
+// Transfers all tokens in the Stride-side holding address over to the communityPoolReturnAddress ICA
+func (k Keeper) IBCReturnAllCommunityPoolTokens(ctx sdk.Context, communityPoolHostZone types.HostZone) error {
+	// Use bankKeeper to see all coin types and amounts currently in the stride-side holding address
+	req := banktypes.NewQueryAllBalancesRequest(sdk.AccAddress(communityPoolHostZone.CommunityPoolHoldingAddress), nil)
+	resp, err := k.bankKeeper.AllBalances(ctx, req)
+	if err != nil {
+		return err
+	}
+
+	memo := ""
+	errors := make([]error, 0)
```

Comment:
rather than using an errors array, I would recommend making a sub-function for the transfer of a given token and then this outer function is something like:

```go
for _, coin := range resp.Balances {
   if err := k.Subfunction(...); err != nil {
      k.Logger(ctx).Error(err.Wrapf("error submitting ibc transfer for %v %s", coin.Amount, coin.Denom).Error())
      continue
  }
}
```

And then the only error returned to this outer function is from [this](https://github.com/Stride-Labs/stride/pull/929/files#diff-dbc8d2b8a50b35a2b45a2483916e27d4256f669d24fa8a0e69ec4378dd0f1771R63). That way, one bad denom doesn't cause this whole function to error

Reply:
Yep good point.  This subfunction wrap suggestion is even more pertinent now because there is more complicated logic around registering a callback and saving args.  I was only doing the errors array for the purpose that a single failure wouldn't stop the rest of the attempted sends but that should still be possible with your method as well.  I am still going through more generally and fixing the errors to use errormod and only log at the top level instead of all through the call stack